pkg/controller/pkg/revision/establisher.go

```diff
@@ -76,84 +64,71 @@ type currentDesired struct {
-	// Reset all objects and references.
-	e.allObjs = nil
-	e.resourceRefs = nil
+// Establish checks that control or ownership of resources can be established by
+// parent, then establishes it.
+func (e *APIEstablisher) Establish(ctx context.Context, objs []runtime.Object, parent resource.Object, control bool) ([]runtimev1alpha1.TypedReference, error) { // nolint:gocyclo
+	allObjs := []currentDesired{}
+	resourceRefs := []runtimev1alpha1.TypedReference{}
 	for _, res := range objs {
 		// Assert desired object to resource.Object so that we can access its
 		// metadata.
 		d, ok := res.(resource.Object)
 		if !ok {
-			return errors.New(errAssertObj)
+			return nil, errors.New(errAssertObj)
 		}
 
 		// Make a copy of the desired object to be populated with existing
 		// object, if it exists.
 		current := res.DeepCopyObject()
 		err := e.client.Get(ctx, types.NamespacedName{Name: d.GetName(), Namespace: d.GetNamespace()}, current)
 		if resource.IgnoreNotFound(err) != nil {
-			return err
+			return nil, err
 		}
 
 		// If resource does not already exist, we must attempt to dry run create
 		// it.
 		if kerrors.IsNotFound(err) {
 			// Add to objects as not existing.
-			e.allObjs = append(e.allObjs, currentDesired{
+			allObjs = append(allObjs, currentDesired{
 				Desired: d,
 				Current: nil,
 				Exists:  false,
 			})
 			if err := e.create(ctx, d, parent, control, client.DryRunAll); err != nil {
-				return err
+				return nil, err
 			}
 			continue
 		}
 
 		c := current.(resource.Object)
 		// Add to objects as existing.
-		e.allObjs = append(e.allObjs, currentDesired{
+		allObjs = append(allObjs, currentDesired{
 			Desired: d,
 			Current: c,
 			Exists:  true,
 		})
 
 		if err := e.update(ctx, c, d, parent, control, client.DryRunAll); err != nil {
-			return err
+			return nil, err
 		}
 	}
 
-	return nil
-}
-
-// Establish establishes control or ownership of resources in the package that
-// have been previously checked as ownable or controllable.
-func (e *APIEstablisher) Establish(ctx context.Context, parent resource.Object, control bool) error {
-	for _, cd := range e.allObjs {
+	for _, cd := range allObjs {
 		if !cd.Exists {
 			if err := e.create(ctx, cd.Desired, parent, control); err != nil {
-				return err
+				return nil, err
 			}
-			e.resourceRefs = append(e.resourceRefs, *meta.TypedReferenceTo(cd.Desired, cd.Desired.GetObjectKind().GroupVersionKind()))
+			resourceRefs = append(resourceRefs, *meta.TypedReferenceTo(cd.Desired, cd.Desired.GetObjectKind().GroupVersionKind()))
 			continue
 		}
 
 		if err := e.update(ctx, cd.Current, cd.Desired, parent, control); err != nil {
-			return err
+			return nil, err
 		}
-		e.resourceRefs = append(e.resourceRefs, *meta.TypedReferenceTo(cd.Desired, cd.Desired.GetObjectKind().GroupVersionKind()))
+		resourceRefs = append(resourceRefs, *meta.TypedReferenceTo(cd.Desired, cd.Desired.GetObjectKind().GroupVersionKind()))
 	}
-	return nil
+
+	return resourceRefs, nil
```

Comment:
I opted to only return resource refs if no errors occur. This doesn't change anything in practice because we don't set the resource refs in the reconciler if an error occurred anyway, but I would prefer to remove the possibility that any consumer of this method sets an incomplete list of resources in the event of an error.